common/kvstore/tablestore/table_store_wrapper.go

```diff
@@ -0,0 +1,353 @@
+package tablestore
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"github.com/Layr-Labs/eigenda/common/kvstore"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"math"
+	"sort"
+)
+
+// Table ID 0 is reserved for use internal use by the metadata table.
+const metadataTableID uint32 = math.MaxUint32
+
+// Table ID 1 is reserved for use by the namespace table. This stores a mapping between IDs and table names.
+const namespaceTableID uint32 = math.MaxUint32 - 1
+
+// The number of tables reserved for internal use.
+const reservedTableCount uint32 = 2
+
+// The first table ID that can be used by user tables.
+const maxUserTableCount = math.MaxUint32 - reservedTableCount
+
+// This key is used to store the schema version in the metadata table.
+const metadataSchemaVersionKey = "schema_version"
+
+// The metadataDeletionKey holds the value of the table name that is being deleted if one is currently being deleted.
+// This is required for atomic deletion of tables.
+const metadataDeletionKey = "deletion"
+
+// The current schema version of the metadata table.
+const currentSchemaVersion uint64 = 0
+
+var _ kvstore.TableStore = &tableStore{}
+
+// ERR_TABLE_LIMIT_EXCEEDED is returned when the maximum number of tables has been reached.
+var ERR_TABLE_LIMIT_EXCEEDED = errors.New("table limit exceeded")
+
+// tableStore is an implementation of TableStore that wraps a Store.
+type tableStore struct {
+	logger logging.Logger
+
+	// A base store implementation that this TableStore wraps.
+	base kvstore.Store
+
+	// A map from table names to table IDs.
+	tableIDMap map[string]uint32
+
+	// A map from table IDs to tables.
+	tableMap map[uint32]kvstore.Table
+
+	// The highest ID of all user tables in the store. Is -1 if there are no user tables.
+	highestTableID int64
+
+	// Builds keys for the metadata table.
+	metadataTable kvstore.Table
+
+	// Builds keys for the namespace table.
+	namespaceTable kvstore.Table
+}
+
+// Wrapper wraps the given Store to create a TableStore.
+//
+// WARNING: it is not safe to access the wrapped store directly while the TableStore is in use. The TableStore uses
+// special key formatting, and direct access to the wrapped store may violate the TableStore's invariants, resulting
+// in undefined behavior.
+func Wrapper(logger logging.Logger, base kvstore.Store) (kvstore.TableStore, error) {
```

Comment:
I'd create a TableStore, instead of asking for an external Store and wrapping it. This way the Store can be internally created and no risk of breaking invariants by accessing it directly outside the TableStore.

Reply:
The purpose of requiring the caller to provide the `Store` implementation is to support multiple `Store` implementations. Currently this is used in unit tests, where I sometimes use a store implementation that is built on top of an in memory map. Let's discuss this.

Reply:
Changed based on our slack discussion.